File: src/main/java/org/veupathdb/lib/blast/BlastQueryConfig.java
package org.veupathdb.lib.blast;

import org.veupathdb.lib.blast.consts.Flag;
import org.veupathdb.lib.blast.field.ExpectValue;
import org.veupathdb.lib.blast.field.Location;
import org.veupathdb.lib.blast.field.QueryFile;
import org.veupathdb.lib.blast.util.JSONObjectDecoder;
import org.veupathdb.lib.blast.util.JSONObjectEncoder;

public interface BlastQueryConfig extends BlastConfig
{
  QueryFile getQueryFile();

  void setQueryFile(String file);

  Location getQueryLocation();

  void setQueryLocation(Location val);

  String getDBFile();

  void setDBFile(String file);

  ExpectValue getExpectValue();

  void setExpectValue(String val);

  Boolean getSoftMasking();

  void setSoftMasking(Boolean val);

  Boolean getLowercaseMasking();

  void setLowercaseMasking(Boolean val);

  String getEntrezQuery();

  void setEntrezQuery(String val);

  Double getQueryCoverageHSPPercent();

  void setQueryCoverageHSPPercent(Double val);

  Long getMaxHSPs();

  void setMaxHSPs(Long val);

  Byte getDBSize();

  void setDBSize(Byte val);

  Short getSearchSpace();

  void setSearchSpace(Short val);

  String getImportSearchStrategy();

  void setImportSearchStrategy(String file);

  String getExportSearchStrategy();

  void setExportSearchStrategy(String file);

  Double getExtensionDropoffUngapped();

  void setExtensionDropoffUngapped(Double val);

  Long getWindowSize();

  void setWindowSize(Long val);

  Boolean getRemote();

  void setRemote(Boolean val);

  @Override
  default void decodeJSON(JSONObjectDecoder node) {
    BlastConfig.super.decodeJSON(node);

    node.decode(Flag.QueryFile, this::setQueryFile);
    node.decode(Flag.QueryLocation, this::setQueryLocation, Location::fromJSON);
    node.decode(Flag.DBFile, this::setDBFile);
    node.decode(Flag.ExpectValue, this::setExpectValue);
    node.decode(Flag.SoftMasking, this::setSoftMasking);
    node.decode(Flag.LowercaseMasking, this::setLowercaseMasking);
    node.decode(Flag.EntrezQuery, this::setEntrezQuery);
    node.decode(Flag.QueryCoverageHSPPercent, this::setQueryCoverageHSPPercent);
    node.decode(Flag.MaxHSPs, this::setMaxHSPs);
    node.decode(Flag.DBSize, this::setDBSize);
    node.decode(Flag.SearchSpace, this::setSearchSpace);
    node.decode(Flag.ImportSearchStrategy, this::setImportSearchStrategy);
    node.decode(Flag.ExportSearchStrategy, this::setExportSearchStrategy);
    node.decode(Flag.ExtensionDropoffUngapped, this::setExtensionDropoffUngapped);
    node.decode(Flag.WindowSize, this::setWindowSize);
    node.decode(Flag.Remote, this::setRemote);
  }

  @Override
  default JSONObjectEncoder toJSON(boolean includeTool) {
    final var out = BlastConfig.super.toJSON(includeTool);

    out.encode(Flag.QueryFile, getQueryFile());
    out.encode(Flag.QueryLocation, getQueryLocation());
    out.encode(Flag.DBFile, getDBFile());
    out.encode(Flag.ExpectValue, getExpectValue());
    out.encode(Flag.SoftMasking, getSoftMasking());
    out.encode(Flag.LowercaseMasking, getLowercaseMasking());
    out.encode(Flag.EntrezQuery, getEntrezQuery());
    out.encode(Flag.QueryCoverageHSPPercent, getQueryCoverageHSPPercent());
    out.encode(Flag.MaxHSPs, getMaxHSPs());
    out.encode(Flag.DBSize, getDBSize());
    out.encode(Flag.SearchSpace, getSearchSpace());
    out.encode(Flag.ImportSearchStrategy, getImportSearchStrategy());
    out.encode(Flag.ExportSearchStrategy, getExportSearchStrategy());
    out.encode(Flag.ExtensionDropoffUngapped, getExtensionDropoffUngapped());
    out.encode(Flag.WindowSize, getWindowSize());
    out.encode(Flag.Remote, getRemote());

    return out;
  }
}

File: src/main/java/org/veupathdb/lib/blast/BlastQueryConfig.kt
package org.veupathdb.lib.blast

import org.veupathdb.lib.blast.consts.Flag
import org.veupathdb.lib.blast.field.QueryFile
import org.veupathdb.lib.blast.field.ExpectValue
import org.veupathdb.lib.blast.field.Location
import org.veupathdb.lib.blast.util.JSONObjectDecoder
import org.veupathdb.lib.blast.util.JSONObjectEncoder

interface BlastQueryConfig : BlastConfig {
  var queryFile: QueryFile?
  var queryLocation: Location?
  var dbFile: String?
  var expectValue: ExpectValue?
  var softMasking: Boolean?
  var lowercaseMasking: Boolean?
  var entrezQuery: String?
  var queryCoverageHSPPercent: Double?
  var maxHSPs: Long?
  var dbSize: Byte?
  var searchSpace: Short?
  var importSearchStrategy: String?
  var exportSearchStrategy: String?
  var extensionDropoffUngapped: Double?
  var windowSize: Long?
  var remote: Boolean?

  override fun decodeJSON(node: JSONObjectDecoder) {
    super.decodeJSON(node)
    node.decodeString(Flag.QueryFile)                { queryFile = QueryFile(it)     }
    node.decodeJSON(Flag.QueryLocation)              { Location(it)                  }
    node.decodeString(Flag.DBFile)                   { dbFile = it                   }
    node.decodeString(Flag.ExpectValue)              { expectValue = ExpectValue(it) }
    node.decodeBool(Flag.SoftMasking)                { softMasking = it              }
    node.decodeBool(Flag.LowercaseMasking)           { lowercaseMasking = it         }
    node.decodeString(Flag.EntrezQuery)              { entrezQuery = it              }
    node.decodeDouble(Flag.QueryCoverageHSPPercent)  { queryCoverageHSPPercent = it  }
    node.decodeLong(Flag.MaxHSPs)                    { maxHSPs = it                  }
    node.decodeByte(Flag.DBSize)                     { dbSize = it                   }
    node.decodeShort(Flag.SearchSpace)               { searchSpace = it              }
    node.decodeString(Flag.ImportSearchStrategy)     { importSearchStrategy = it     }
    node.decodeString(Flag.ExportSearchStrategy)     { exportSearchStrategy = it     }
    node.decodeDouble(Flag.ExtensionDropoffUngapped) { extensionDropoffUngapped = it }
    node.decodeLong(Flag.WindowSize)                 { windowSize = it               }
    node.decodeBool(Flag.Remote)                     { remote = it                   }
  }

  override fun toJSON(includeTool: Boolean): JSONObjectEncoder {
    val out: JSONObjectEncoder = super.toJSON(includeTool)
    out.encode(Flag.QueryFile, queryFile)
    out.encode(Flag.QueryLocation, queryLocation)
    out.encode(Flag.DBFile, dbFile)
    out.encode(Flag.ExpectValue, expectValue)
    out.encode(Flag.SoftMasking, softMasking)
    out.encode(Flag.LowercaseMasking, lowercaseMasking)
    out.encode(Flag.EntrezQuery, entrezQuery)
    out.encode(Flag.QueryCoverageHSPPercent, queryCoverageHSPPercent)
    out.encode(Flag.MaxHSPs, maxHSPs)
    out.encode(Flag.DBSize, dbSize)
    out.encode(Flag.SearchSpace, searchSpace)
    out.encode(Flag.ImportSearchStrategy, importSearchStrategy)
    out.encode(Flag.ExportSearchStrategy, exportSearchStrategy)
    out.encode(Flag.ExtensionDropoffUngapped, extensionDropoffUngapped)
    out.encode(Flag.WindowSize, windowSize)
    out.encode(Flag.Remote, remote)
    return out
  }
}